Detour/Source/DetourNavMesh.cpp

```diff
@@ -1520,3 +1529,51 @@ dtStatus dtNavMesh::getPolyArea(dtPolyRef ref, unsigned char* resultArea) const
 	return DT_SUCCESS;
 }
 
+#ifdef NAVMESH_ISLAND_SYSTEM
+dtStatus dtNavMesh::setPolyIslandIdx(dtPolyRef ref, unsigned int idx)
+{
+	if (!ref) return DT_FAILURE;
+	unsigned int salt, it, ip;
+	decodePolyId(ref, salt, it, ip);
+	if (it >= (unsigned int)m_maxTiles) return DT_FAILURE | DT_INVALID_PARAM;
+	if (m_tiles[it].salt != salt || m_tiles[it].header == 0) return DT_FAILURE | DT_INVALID_PARAM;
+	dtMeshTile* tile = &m_tiles[it];
+	if (ip >= (unsigned int)tile->header->polyCount) return DT_FAILURE | DT_INVALID_PARAM;
+	dtPoly* poly = &tile->polys[ip];
+
+	poly->setIslandIdx(idx);
+
+	return DT_SUCCESS;
+}
+
+dtStatus dtNavMesh::getPolyIslandIdx(dtPolyRef ref, unsigned int* idx) const
+{
+	if (!ref || !idx) return DT_FAILURE;
+	unsigned int salt, it, ip;
+	decodePolyId(ref, salt, it, ip);
+	if (it >= (unsigned int)m_maxTiles) return DT_FAILURE | DT_INVALID_PARAM;
+	if (m_tiles[it].salt != salt || m_tiles[it].header == 0) return DT_FAILURE | DT_INVALID_PARAM;
+	const dtMeshTile* tile = &m_tiles[it];
+	if (ip >= (unsigned int)tile->header->polyCount) return DT_FAILURE | DT_INVALID_PARAM;
+	const dtPoly* poly = &tile->polys[ip];
+
+	*idx = poly->getIslandIdx();
+
+	return DT_SUCCESS;
+}
+
+bool dtNavMesh::checkPathExists(const dtPolyRef startRef, const dtPolyRef endRef) const
+{
+	unsigned int islandIndexFrom = 0;
+	unsigned int islandIndexTo = 0;
+	if (dtStatusFailed(getPolyIslandIdx(startRef, &islandIndexFrom)))
+		return false;
+	if (dtStatusFailed(getPolyIslandIdx(endRef, &islandIndexTo)))
+		return false;
+	if (islandIndexFrom == 0 || islandIndexTo == 0)
+		return false;
+	if (islandIndexFrom != islandIndexTo)
+		return false;
+	return true;
+}
```

Comment:
I have not checked your implementation closely, but I don't see how this can be correct. This will give symmetric answers, i.e. `checkPathExists(a, b) == checkPathExists(b, a)`. But that is certainly not correct in the fact of unidirectional offmesh connections.

Reply:
My system not work with offmesh connections.

Reply:
I think I will do this system with offmesh connections, but later/